### ISSUE TITLE
On effects, this.reducer() is not correctly infered

### ISSUE THREAD

**Issue by semoal** [rematch/rematch#870]
**Describe the bug**

With Rematch you can use `this` inside effects to dispatch data to reducers.

Basically this:

```tsx
export const players = createModel<RootModel>()({
	state: {
		players: [],
	},
	reducers: {
		SET_PLAYERS: (state, players) => {
			return {
				...state,
				players,
			}
		},
	},
	effects: () => ({
    async getPlayers() {
      const response = await fetch(
        'https://www.balldontlie.io/api/v1/players'
      )
      const { data } = await response.json()
      this.SET_PLAYERS(data)
    }
  })
})
```

is equivalent to this:

```tsx
export const players = createModel<RootModel>()({
	state: {
		players: [],
	},
	reducers: {
		SET_PLAYERS: (state, players) => {
			return {
				...state,
				players,
			}
		},
	},
	effects: (dispatch) => ({
    async getPlayers() {
      const response = await fetch(
        'https://www.balldontlie.io/api/v1/players'
      )
      const { data } = await response.json()
      dispatch.players.SET_PLAYERS(data)
    }
  })
})
```

**To Reproduce**

When using Typescript, this.SET_PLAYERS() won't get autocompleted because this isn't overriden by Rematch.

**Current work around**

```
      (this as typeof players).SET_PLAYERS(data)

````

**Expected behavior**

Should autocomplete reducers and effects of the current Model.

**Additional context**

- Version: 2.0.1

**Comment by maciekgrzybek** [comment]
I'm happy to take a look at this :)

**Comment by maciekgrzybek** [comment]
It's really tough to figure it out really... Do you have any ideas what might be causing that?

**Comment by semoal** [comment]
Actually this type:
```ts
export type ModelEffect<
	TModels extends Models<TModels> = Record<string, any>
> = (
	payload: Action['payload'],
	rootState: RematchRootState<TModels>,
	meta: Action['meta']
) => any
```

Is the current type for an effect, you will see that if you add the `this` property with anything else, you get autocomplete of the rootstate:

```ts
export type ModelEffect<
	TModels extends Models<TModels> = Record<string, any>
> = (
    this: RematchRootState<TModels>
	payload: Action['payload'],
	rootState: RematchRootState<TModels>,
	meta: Action['meta']
) => any
```

We must add something like: `ExtractReducers&Effects<TModel>` where we can pass a Model and extract his effects and reducers. 
 
It is not an easy fix, @tianzhich it's an expert with types and made something impossible to practically any library out there, is that being practically 100% compatible with the current Redux and React Redux typings.

**Comment by maciekgrzybek** [comment]
Awesome, thanks for explanation, I'll give it another go :)

**Comment by maciekgrzybek** [comment]
I have another question :) I'm not sure if I'm using the correct local examples to actually confirm my changes in the core? is it just the `examples` folder? Is it linked automatically to the local rematch? Unfortunately `CONTRIBUTING.md` file is not really specific about it :(

**Comment by semoal** [comment]
Yes the examples folder is linked via the monorepo. 
You only need to re-run the build on the package you change.

```
yarn build --scope @rematch/core
```
In the root of the project and will rebuild in less than 5 seconds the package =)

**Comment by semoal** [comment]
We're using `tsdx` internally for the packages, so probably you could use also `tsdx watch` inside the core package (not sure about this)
Tomorrow I'll improve the CONTRIBUTION.md =)

**Comment by maciekgrzybek** [comment]
Wow, that was a quick response :D Thanks a lot :)

**Comment by tianzhich** [comment]
Hi mate, how about the progress of this? @maciekgrzybek I can continue from your work if you did some modifications.

**Comment by maciekgrzybek** [comment]
Hey @tianzhich I totally dropped the ball on this one, I had few approaches but couldn't figure it out at all :(

**Comment by tianzhich** [comment]
I created a [minimal rematch typing system](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgLIHsAmEA2AeAFWQgA9IRMBnNLXSwgPgeQG8AoZT5KCTAVyRRKALmQAlXgOiUA3By4QYMCAjAjkAUSUrI9AgzkBfNmzABPAA4oJ-QdQC84leiiY8lMFFABzADRPbaAMTcytNbVUIPWIyCApqDGwcPQZ-MRjyKgCpIWRHCQQXNw8vED9kEAgANyDmfOdXd08ffy1lVUI0pmQAH3D2sABhHjgwF07xYLZQSFhEFES6Qgy4rMXkxmZ2LmQAbQBrCDNREp8AXVF1xiMQy2tJQTzkAAoASjzmKvRgTDlTO-6OmWpEyCVoG30aRW8WydieBSKTVK5UqNSg3UczzAAAtgOoNGQoIgwAAxKDoAC2Nhy9DEDHe9k+31+tzCbR0wwgo3GRBBqzBSRSUL5MOpcPqhUapzK-lRtSez0weIsowQ2NEBM8xLJlPWKQZzARUuaMsBHQIXQYrJQmqJqh1FL1wNiMKd+jqrHknAOR2QoGQhzM6BgyAIF00hO15Md4ImBB9ZjOwWM-zZkft0auvJda3BQuQAFVobmkpsnrao5SxdI8AXdgByHiBIT1pPIABkEa1Gcp7NUen8BatqZt6dJ0erQjw6RFWUnlA92y4Cb9IADR2D4nDYgTZxuI67dvHvYiamdoJogsYg+LAtwZccdfrigGlFbt5eSsoKrAao1Y4dN16Q+VdlCgcRkAAfgg2dqHnKCINEOVwNEJ8Xx0N89xMGA+BAVRgHQNcEBGSAsw-ICGDeT0dh4MA+CgNdpw-edWg-PtdAmOlKIpdBRCXHZOCbGlRDEGRiFPfE5B2QwDWogTBIgOiGOQHipK4Qx90KEAPBU8EAEYnmIrlSNjMR0HQMB1ko15nn47gHmkPivQEmBzL0qiWGQQx-AAIzgKB3PeTzjGk3xnPQ-tREVZVVWxWTbOQPyAC9AtYLyvNeNgZJMLSdJ4pIACZDJIiArjMiyrLeWznKEuwnPkzhXPQAqPK83z-JaoKvOc7zwokqKvx-NV4rsnYAHoACoJucgSJuQAAFHAuUoFAKTgQ5kDgHAcGQHEUBmXAcGAbw4iQagAHcXH2ABCGadgmsa7s4ZLOrkhquDGsbkAQOAVr0k4+AsCwXDAZBBti8SBk2-DCOQGBo10pI9KegTwd-bEADp8twPSMeSwK-ne8avp+laCoBoGQbBmL0YqFw1p24kCLXeHKURnGUZ2NG1Sx-SMaagmTCJj6Sd+iAAGYKeBqBQe57FIZ0aGwGZuGEYQeieHAdmcGebGcAKzLhc4OXecKvG4CSzrCaNz7vrFgAWKWqblumoAZpWVdZilvo1uJQb13XwQNzmuBNvWCv58yraFm3RZWgBWJ2ZeQAADHE8RT3bLBaVP08oTPgZmXb0F27EUHVqBNf98Fqe-WLoBDzg88j5q3jU4Xm5etvG+QHvbfO7EzFTpqCszn6tZ8-aQDAptvpcHhVBwMxIJ7sOg5b6OjdLvEN+796QvUzLsumcBoHgJBxHMyya9gy8lnK6+ki2Zy9f+rOrE3V+wp2cPRFCCBP5ByytaZAABBcsAFozsVpFfdY-h-6AMKsOIAA) and tried to fix this but failed. Maybe someone can pick this one and continue from here.

**Comment by semoal** [comment]
In the new release we're going to ship (2.1.0): `this.REDUCER` will be typed as:
```
 () => (payload?: any, meta?: any)
```
So `this.reducer` won't never fail again, but if you want to type it in a more safe way you can:

```ts
		const count: CountModel = {
			state: 0,
			reducers: {
				addOne: (s: number): number => s + 1,
			},
			effects: {
				incrementOneThroughEffect(this: ModelDispatcher<CountModel, RootModel>): void {
					this.addOne()
				},
			},
		}
```

**Comment by semoal** [comment]
Close this, since will be fixed with 2.1.0 release

**Comment by semoal** [comment]
It is partially, we couldn’t fix a 100% type checking it shouldn’t return error at least. Contributions are appreciated ;)
